### PR TITLE
Add FAQ chapter to CONTRIBUTING.md

### PR DESCRIPTION
**Pull request by JohannesPtaszyk**
Added FAQ with questions I was asking myself, when I first opened the project in Android Studio.

Signed-off-by: Johannes Ptaszyk <johannes@ptaszyk.de>

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -165,3 +165,27 @@ general guidelines for the community as a whole:
   respond to an email you are potentially sending to a large number of
   people.  Please consider this before you update.  Also remember that
   nobody likes spam.
+
+## FAQ
+
+**Is it fine to go with kotlin for my contribution?**
```
Comment (mueller-ma) [inline CONTRIBUTING.md]: s/kotlin/Kotlin :)
Nothing against Kotline from my side. https://github.com/openhab/openhab-android/issues/428

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -165,3 +165,27 @@ general guidelines for the community as a whole:
...
+
+**Should my code be tested? If yes to which degree?**
+
+**Is there a way I can test all the features against a test server or something similar?**
+
+**Do we have guidelines for code style?**
+
+**How does the CD/CI pipeline work?**
+
+**Do I have to include Java doc/ KDoc?**
```
Comment (mueller-ma) [inline CONTRIBUTING.md]: IMO you don't have to if your function name and parameter names are self describing. For complex functions a short Javadoc is appreciated.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -165,3 +165,27 @@ general guidelines for the community as a whole:
...
+
+**How is navigation managed?**
+
+**Do we use fragments in our activities? If yes, are they necessary?**
+
+**Should my code be tested? If yes to which degree?**
+
+**Is there a way I can test all the features against a test server or something similar?**
+
+**Do we have guidelines for code style?**
```
Comment (mueller-ma) [inline CONTRIBUTING.md]: Yes, Google Java Style Guide: https://google.github.io/styleguide/javaguide.html

File: CONTRIBUTING.md (hunk 1 above)
Comment (maniac103) [reply CONTRIBUTING.md]: For the reasons given in https://github.com/openhab/openhab-android/issues/428#issuecomment-340157780 I think I'd prefer to continue using Java though, unless there's a compelling argument for using Kotlin in that particular case.
I don't have a strong opinion on Kotlin vs. Java per she, but I really don't like mixing both languages.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -165,3 +165,27 @@ general guidelines for the community as a whole:
...
+
+**Where does the sitemap rendering happen?**
+
+**How is navigation managed?**
+
+**Do we use fragments in our activities? If yes, are they necessary?**
+
+**Should my code be tested? If yes to which degree?**
+
+**Is there a way I can test all the features against a test server or something similar?**
```
Comment (maniac103) [inline CONTRIBUTING.md]: Either test against the demo server, or setup an own server. For the latter case, it's probably useful to do a checkout of the server side code an run that, following the procedure in https://www.openhab.org/docs/developer/development/ide.html

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -165,3 +165,27 @@ general guidelines for the community as a whole:
...
+
+**Is there an architectural design for the data layer?**
+
+**Where does the sitemap rendering happen?**
+
+**How is navigation managed?**
+
+**Do we use fragments in our activities? If yes, are they necessary?**
+
+**Should my code be tested? If yes to which degree?**
```
Comment (maniac103) [inline CONTRIBUTING.md]: Ideally yes, to whatever extent is doable and meaningful. It's not a requirement though.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -165,3 +165,27 @@ general guidelines for the community as a whole:
...
+
+**Is there an architectural design for the presentation layer?**
+
+**Is there an architectural design for the data layer?**
+
+**Where does the sitemap rendering happen?**
+
+**How is navigation managed?**
+
+**Do we use fragments in our activities? If yes, are they necessary?**
```
Comment (maniac103) [inline CONTRIBUTING.md]: The answer to the first question is 'it depends', but I don't understand the aim of the latter question. They should be used whenever it makes sense to use them.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -165,3 +165,27 @@ general guidelines for the community as a whole:
...
+
+## FAQ
+
+**Is it fine to go with kotlin for my contribution?**
+
+**Is there an architectural design for the presentation layer?**
+
+**Is there an architectural design for the data layer?**
+
+**Where does the sitemap rendering happen?**
```
Comment (maniac103) [inline CONTRIBUTING.md]: WidgetListFragment + (especially) WidgetAdapter

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -165,3 +165,27 @@ general guidelines for the community as a whole:
...
+
+**How is navigation managed?**
```
Comment (maniac103) [inline CONTRIBUTING.md]: Please clarify that question; I'm not sure what the actual question is here.

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -165,3 +165,27 @@ general guidelines for the community as a whole:
...
   people.  Please consider this before you update.  Also remember that
   nobody likes spam.
+
+## FAQ
+
+**Is it fine to go with kotlin for my contribution?**
+
+**Is there an architectural design for the presentation layer?**
+
+**Is there an architectural design for the data layer?**
```
Comment (maniac103) [inline CONTRIBUTING.md]: Design as in design documentation? No, there isn't.
Design as in 'meaningful structuring of code' We did our best to achieve that ;-)

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -165,3 +165,27 @@ general guidelines for the community as a whole:
...
+
+**Do we use fragments in our activities? If yes, are they necessary?**
+
+**Should my code be tested? If yes to which degree?**
+
+**Is there a way I can test all the features against a test server or something similar?**
+
+**Do we have guidelines for code style?**
+
+**How does the CD/CI pipeline work?**
```
Comment (mueller-ma) [inline CONTRIBUTING.md]: After a PR is created, Travis CI builds the app and runs the full testsuite. Must pass to get merged.
Cirrus CI validates the strings (e.g. correct spelling of "openHAB"). Must pass to get merged.
Stickler CI runs Checkstyle. Sometime the checkstyle issue are false positives, so this needn't pass to get merged.
DCO checks if the commits are signed. Must pass to get merged, except for very small changes.

File: CONTRIBUTING.md (hunk 4 above)
Comment (mueller-ma) [reply CONTRIBUTING.md]: If you don't want to deploy a second server, you can create a sitemap for testing the app.

File: CONTRIBUTING.md (hunk 1 above)
Comment (JohannesPtaszyk) [reply CONTRIBUTING.md]: Hmm, so we should kinda aggree on one of the answers her. Kotlin seems to be growing real quick, but I still see your argument as valid. Java is spread world wide while kotlin probably is only in the Android community.

File: CONTRIBUTING.md (hunk 9 above)
Comment (JohannesPtaszyk) [reply CONTRIBUTING.md]: Design as 'meaningful structuring of code' + use of designpatterns like mvvm/mvp/mvc for the ui layer or repository pattern, use cases for the data or domain layer.

File: CONTRIBUTING.md (hunk 8 above)
Comment (JohannesPtaszyk) [reply CONTRIBUTING.md]: Is navigation managed in one place? Like for example the new architecture navigation component.

Comment (maniac103) [reply CONTRIBUTING.md]: No. ContentController manages how fragments are displayed, and MainActivity commands it what fragments to display. Navigation is too dynamic in our app (determined by sitemap) for the navigation architecture component to be useful.

File: CONTRIBUTING.md (hunk 9 above)
Comment (maniac103) [reply CONTRIBUTING.md]: We have proper separation between data model (see model package) and presentation (see ui package). We don't separate matters further, so no mvvm/mvp/mvc. As the application isn't overly complex I think this is fine.

File: CONTRIBUTING.md (hunk 1 above)
Comment (maniac103) [reply CONTRIBUTING.md]: > Java is spread world wide while kotlin probably is only in the Android community.

Exactly. In our specific case the server side is written in Java as well, so it's helpful to use the same language in the client.

**Comment by JohannesPtaszyk** [comment]
I am currently busy, but I'll add your answers to the document and commit it as soon as possible. I will probably not use your exact words, but rather just write a short sentence for each question. Thnk you for the answers!

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -168,24 +168,57 @@ general guidelines for the community as a whole:
...
+For the latter case, it's probably useful to do a checkout of the server side code an run that,
+following the procedure in https://www.openhab.org/docs/developer/development/ide.html
+
 **Do we have guidelines for code style?**
 
+Yes, Google Java Style Guide: https://google.github.io/styleguide/javaguide.html
+
 **How does the CD/CI pipeline work?**
 
-**Do I have to include Java doc/ KDoc?**
```
Comment (mueller-ma) [inline CONTRIBUTING.md]: Nicely formatted table:

| Service | Shown as | Task | Desired result | Howto fix |
|--|--|--|--|--|
| Travis CI | continuous-integration/travis-ci/(pr,push) | Runs full build and test suite | Must pass | Look at the build output and try to run the test locally. If the tests are failing even if you haven't changed the particular code, please poke @openhab/android-maintainers and ask them to rerun the job. |
| Stickler CI | stickler-ci | Runs checkstyle to ensure the Google Java Style Guide is followed | Should pass, but sometimes it reports false-positives | Look at the comments (https://github.com/openhab/openhab-android/pull/\<pr-number\>/files) |
| Cirrus CI | Validate strings | Checks for common mistakes in strings.xml, e.g. wrong spelling of "openHAB" | Must pass | Look at the checks output and fix the issues. |
| Developer Certificate of Origin | DCO | Checks if commits are signed-of-by | Must pass | `git commit -s --amed` and `git push -f` |
| WIP | WIP | Checks if the PR name contains "WIP" (Work in progress) | Must pass | Remove "WIP" from the PR title when you're done. |

Comment (JohannesPtaszyk) [reply CONTRIBUTING.md]: Will add it!

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -165,3 +165,65 @@ general guidelines for the community as a whole:
   respond to an email you are potentially sending to a large number of
   people.  Please consider this before you update.  Also remember that
   nobody likes spam.
+
+## FAQ
+
+**Is it fine to go with Kotlin for my contribution?**
+
+We don't use Kotlin. This is due to the fact, that our main product is written in pure java.
```
Comment (mueller-ma) [inline CONTRIBUTING.md]: "... that the **server components** are written in pure **Java**".

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -165,3 +165,65 @@ general guidelines for the community as a whole:
...
+Even tough the Android community is adapting Kotlin well, most guys from the Java community
+have never seen Kotlin code before.
```
Comment (mueller-ma) [inline CONTRIBUTING.md]: I don't like the phrase "have never seen Kotlin code before". I guess most of them have seen Kotlin code, but aren't used to it.

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -165,3 +165,65 @@ general guidelines for the community as a whole:
...
+
+**Is there an architectural design for the presentation layer?**
+
+There is no MVVM/MVP or MVC in place, because the app is not too complex and it would
```
Comment (mueller-ma) [inline CONTRIBUTING.md]: I suggest to remove linebreaks like this one.

File: CONTRIBUTING.md (hunk 7 above)
Comment (mueller-ma) [reply CONTRIBUTING.md]: "Sitemap"

File: CONTRIBUTING.md [hunk 15]
```diff
@@ -165,3 +165,65 @@ general guidelines for the community as a whole:
...
+The answer to the first question is 'it depends'.
+They should be used whenever it makes sense to use them.
+
+**Should my code be tested? If yes to which degree?**
+
+Ideally yes, to whatever extent is doable and meaningful. It's not a requirement though.
+
+**Is there a way I can test all the features against a test server or something similar?**
+
+Either test against the demo server, or setup an own server.
```
Comment (mueller-ma) [inline CONTRIBUTING.md]: Link "demo server" to https://demo.openhab.org

File: CONTRIBUTING.md [hunk 16]
```diff
@@ -165,3 +165,65 @@ general guidelines for the community as a whole:
...
+For the latter case, it's probably useful to do a checkout of the server side code an run that,
+following the procedure in https://www.openhab.org/docs/developer/development/ide.html
+
+**Do we have guidelines for code style?**
+
+Yes, Google Java Style Guide: https://google.github.io/styleguide/javaguide.html
```
Comment (mueller-ma) [inline CONTRIBUTING.md]: "Also omit "openHAB" or "My" from class names."

File: CONTRIBUTING.md [hunk 17]
```diff
@@ -165,3 +165,65 @@ general guidelines for the community as a whole:
...
+
+**How does the CD/CI pipeline work?**
```
Comment (mueller-ma) [inline CONTRIBUTING.md]: CI/CD is more common: https://en.wikipedia.org/wiki/CI/CD

File: CONTRIBUTING.md [hunk 18]
```diff
@@ -165,3 +165,65 @@ general guidelines for the community as a whole:
...
+For the latter case, it's probably useful to do a checkout of the server side code an run that,
+following the procedure in https://www.openhab.org/docs/developer/development/ide.html
+
+**Do we have guidelines for code style?**
+
+Yes, Google Java Style Guide: https://google.github.io/styleguide/javaguide.html
+
+**How does the CD/CI pipeline work?** 
+
+<br>
```
Comment (mueller-ma) [inline CONTRIBUTING.md]: Is this really needed?

Comment (JohannesPtaszyk) [reply CONTRIBUTING.md]: Github renders this better. I just had the preview in Android Studio up and it looked bad without it. -> Removed it

Comment (Jochen1980) [inline CONTRIBUTING.md]: Please merge this PR that all of us can improve the FAQ section! Thanks in advance.

File: CONTRIBUTING.md [hunk 19]
```diff
@@ -165,3 +165,70 @@ general guidelines for the community as a whole:
...
   nobody likes spam.
+
+## FAQ
+
+**Is it fine to go with Kotlin for my contribution?**
+
+We don't use Kotlin. This is due to the fact, that our server components
+are written in pure Java. Even tough the Android community is adapting
+Kotlin well, most guys from the Java community are not used to writing
+Kotlin code before.
```
Comment (mueller-ma) [inline CONTRIBUTING.md]: The word "before" seems wrong to me here.

File: CONTRIBUTING.md [hunk 20]
```diff
@@ -165,3 +165,70 @@ general guidelines for the community as a whole:
...
+
+ContentController manages how fragments are displayed, and MainActivity 
+commands it what fragments to display. Navigation is too dynamic in our
+app (determined by Sitemap) for the navigation architecture component to
+be useful.
+
+**Do we use fragments in our activities? If yes, are they necessary?**
+
+The answer to the first question is 'it depends'.
+They should be used whenever it makes sense to use them.
```
Comment (mueller-ma) [inline CONTRIBUTING.md]: For very simple activities, e.g. LogActivity, we don't use fragments, otherwise we do.

**Comment by JohannesPtaszyk** [comment]
So I added the things. But I think we should merge soon and give others the chance to improve upon it. :)

**Comment by mueller-ma** [comment]
@JohannesPtaszyk I merged and squashed your commits. Thanks for your contribution. You can now delete your branch locally and from Github.

**Comment by JohannesPtaszyk** [comment]
Thank you!

File: CONTRIBUTING.md (hunk 1 above)
Comment (maniac103) [reply CONTRIBUTING.md]: @mueller-ma Maybe we should revisit this looking at recent news (e.g. https://techcrunch.com/2019/05/07/kotlin-is-now-googles-preferred-language-for-android-app-development/) ... I've toyed around with it for a bit, and initial conversion would be pretty straightforward, with potential of saving about 10% of LoC.

**Pull request merged**